### PR TITLE
[DEV-2797] user status deleted

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/subdomains/generic/user/models/user-data/user-data.entity.ts [hunk 1]
```diff
@@ -319,10 +320,7 @@ export class UserData extends IEntity {
   }
 
   blockUserData(): UpdateResult<UserData> {
-    const update: Partial<UserData> = {
-      status: UserDataStatus.BLOCKED,
-      kycLevel: KycLevel.TERMINATED,
-    };
+    const update: Partial<UserData> = { status: UserDataStatus.DELETED };
```
[the remark this answers is not in the document]
Comment (davidleomay) [reply src/subdomains/generic/user/models/user-data/user-data.entity.ts:323]: Set KYC steps (ident, financial) to cancelled?

File: src/subdomains/generic/user/models/user/user.service.ts [hunk 2]
```diff
@@ -115,7 +115,9 @@ export class UserService {
       .leftJoin('user.userData', 'userData')
       .where('user.refCredit - user.paidRefCredit > 0')
       .andWhere('user.status != :userStatus', { userStatus: UserStatus.BLOCKED })
-      .andWhere('userData.status != :userDataStatus', { userDataStatus: UserDataStatus.BLOCKED })
+      .andWhere('userData.status NOT IN (:...userDataStatus)', {
+        userDataStatus: [UserDataStatus.BLOCKED, UserDataStatus.DELETED],
+      })
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user/user.service.ts:120]: Same for user status?

File: src/subdomains/generic/user/models/user/user.service.ts [hunk 3]
```diff
@@ -239,9 +241,9 @@ export class UserService {
   async blockUser(id: number, allUser = false): Promise<void> {
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user/user.service.ts:241]: Maybe we should rename all the methods as well? Maybe deactivate?

**Review by davidleomay — COMMENTED** [review]
- [ ] Reactivate on merge
- [ ] Update existing users

File: src/subdomains/generic/kyc/dto/mapper/kyc-info.mapper.ts [hunk 4]
```diff
@@ -43,7 +43,7 @@ export class KycInfoMapper {
 
   // --- HELPER METHODS --- //
   private static getUiSteps(userData: UserData): KycStep[] {
-    if (userData.isKycTerminated) return [];
+    if (userData.isBlocked || userData.isDeleted) return [];
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/dto/mapper/kyc-info.mapper.ts:46]: I think we can revert this. Only if the KYC is terminated we want to hide the KYC steps

File: src/subdomains/generic/kyc/enums/kyc.enum.ts [hunk 5]
```diff
@@ -65,6 +65,7 @@ export enum KycStepStatus {
   CANCELED = 'Canceled',
   IGNORED = 'Ignored',
   OUTDATED = 'Outdated',
+  DEACTIVATED = 'Deactivated',
 }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/enums/kyc.enum.ts:69]: I think we can use the `Canceled` status instead. Canceled steps are not shown in UI.

File: src/subdomains/generic/kyc/services/kyc-admin.service.ts [hunk 6]
```diff
@@ -47,6 +47,13 @@ export class KycAdminService {
     await this.kycStepRepo.save(kycStep);
   }
 
+  async deactivateKycSteps(kycSteps: KycStep[]): Promise<void> {
+    for (const kycStep of kycSteps) {
+      if ([KycStepName.FINANCIAL_DATA, KycStepName.IDENT].includes(kycStep.name) && kycStep.isCompleted)
+        await this.kycStepRepo.update(kycStep.id, { status: KycStepStatus.DEACTIVATED });
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc-admin.service.ts:53]: `kycStep.isDone`?

File: src/subdomains/generic/user/models/user-data/user-data.service.ts [hunk 7]
```diff
@@ -291,8 +293,9 @@ export class UserDataService {
     return { user: userData, isKnownUser };
   }
 
-  async blockUserData(userData: UserData): Promise<void> {
+  async deactivateUserData(userData: UserData): Promise<void> {
     await this.userDataRepo.update(...userData.blockUserData());
+    await this.kycAdminService.deactivateKycSteps(userData.kycSteps);
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user-data/user-data.service.ts:298]: `deactivateUserData`?

File: src/subdomains/generic/user/models/user-data/user-data.service.ts [hunk 8]
```diff
@@ -550,6 +553,9 @@ export class UserDataService {
       .copyFiles(slave.id, master.id)
       .catch((e) => this.logger.critical('Error in document copy files:', e));
 
+    if (slave.isDeleted) slave.status = master.status;
+    if (master.isDeleted) master.status = slave.status;
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user-data/user-data.service.ts:557]: Are you sure we should reactivate the slave? It's dead anyway 🤔

File: src/subdomains/generic/user/models/user-data/user-data.entity.ts [hunk 9]
```diff
@@ -320,8 +321,8 @@ export class UserData extends IEntity {
 
   blockUserData(): UpdateResult<UserData> {
     const update: Partial<UserData> = {
-      status: UserDataStatus.BLOCKED,
-      kycLevel: KycLevel.TERMINATED,
+      status: UserDataStatus.DELETED,
+      kycLevel: this.kycLevel < KycLevel.LEVEL_20 ? this.kycLevel : KycLevel.LEVEL_20,
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user-data/user-data.entity.ts:325]: `Math.min`

File: src/subdomains/generic/user/models/user/user.service.ts [hunk 10]
```diff
@@ -236,22 +238,22 @@ export class UserService {
...
-  async blockUser(id: number, allUser = false): Promise<void> {
+  async deactivateUser(id: number, allUser = false): Promise<void> {
     const mainUser = await this.userRepo.findOne({ where: { id }, relations: ['userData', 'userData.users'] });
     if (!mainUser) throw new NotFoundException('User not found');
-    if (mainUser.userData.status === UserDataStatus.BLOCKED)
-      throw new BadRequestException('User Account already blocked');
-    if (mainUser.status === UserStatus.BLOCKED) throw new BadRequestException('User already blocked');
+    if (mainUser.userData.status === UserDataStatus.DELETED)
+      throw new BadRequestException('User Account already deleted');
+    if (mainUser.status === UserStatus.DELETED) throw new BadRequestException('User already deleted');
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user/user.service.ts:246]: Check for Blocked and Deleted?

File: src/subdomains/generic/user/models/user/user.service.ts [hunk 11]
```diff
@@ -114,8 +114,10 @@ export class UserService {
       .createQueryBuilder('user')
       .leftJoin('user.userData', 'userData')
       .where('user.refCredit - user.paidRefCredit > 0')
-      .andWhere('user.status != :userStatus', { userStatus: UserStatus.BLOCKED })
-      .andWhere('userData.status != :userDataStatus', { userDataStatus: UserDataStatus.BLOCKED })
+      .andWhere('user.status NOT IN (:...userStatus)', { userStatus: [UserStatus.BLOCKED, UserStatus.DELETED] })
+      .andWhere('userData.status NOT IN (:...userDataStatus)', {
+        userDataStatus: [UserDataStatus.BLOCKED, UserDataStatus.DELETED],
+      })
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user/user.service.ts:120]: Same change is required in
- UserDtoMapper
- getAllLinkedUsers above
- AuthService
- UserEntity (isPaymentStatusEnabled)

**Comment by TaprootFreak** [comment]
deactivated fits better than delete

File: src/subdomains/generic/user/models/user/user.entity.ts [hunk 12]
```diff
@@ -161,4 +158,8 @@ export class User extends IEntity {
 
     return customChains[this.wallet.name] ?? CryptoService.getBlockchainsBasedOn(this.address);
   }
+
+  get isBlockedOrDeleted(): boolean {
+    return [UserStatus.BLOCKED, UserStatus.DEACTIVATED].includes(this.status);
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user/user.entity.ts:163]: `isBlockedOrDeactivated`?

File: src/subdomains/generic/kyc/services/kyc-admin.service.ts [hunk 13]
```diff
@@ -47,6 +47,13 @@ export class KycAdminService {
     await this.kycStepRepo.save(kycStep);
   }
 
+  async deactivateKycSteps(kycSteps: KycStep[]): Promise<void> {
+    for (const kycStep of kycSteps) {
```
[the remark this answers is not in the document]
Comment (TaprootFreak) [reply src/subdomains/generic/kyc/services/kyc-admin.service.ts]: ich finde resetKyc passt schon, alles bis level 20 ist nur ein formular und nicht wirklich kyc

**Review by davidleomay — APPROVED** [review]

**Pull request merged**